app/waf/src/controllers/application.controller.ts

```diff
@@ -175,9 +197,115 @@ export class ApplicationController {
   }
 
+  private async loadApplication(app: Application): Promise<void> {
+    app.services = await this.applicationRepository.services(app.id).find();
+
+    for (let service of app.services) {
+      await this.loadService(service);
+    }
+  }
+
+  private async loadService(service: Service): Promise<void> {
+    if (service.defaultPoolId) {
+      service.defaultPool = await this.poolRepository.findById(
+        service.defaultPoolId,
+      );
+      await this.loadPool(service.defaultPool);
+    }
+
+    let assocs = await this.serviceEndpointpolicyAssociationRepository.find({
+      where: {
+        serviceId: service.id,
+      },
+    });
+
+    let policyIds = assocs.map(({endpointpolicyId}) => endpointpolicyId);
+    service.policies = await this.endpointpolicyRepository.find({
+      where: {
+        id: {
+          inq: policyIds,
+        },
+      },
+    });
+
+    for (let policy of service.policies) {
+      await this.loadEndpointpolicy(policy);
+    }
+  }
+
+  private async loadPool(pool: Pool): Promise<void> {
+    pool.members = await this.poolRepository.members(pool.id).find();
+
+    for (let member of pool.members) {
+      await this.loadMember(member);
+    }
+
+    let assocs = await this.poolMonitorAssociationRepository.find({
+      where: {
+        poolId: pool.id,
+      },
+    });
+
+    let monitorIds = assocs.map(({monitorId}) => monitorId);
+    pool.monitors = await this.monitorRepository.find({
+      where: {
+        id: {
+          inq: monitorIds,
+        },
+      },
+    });
+  }
+
+  private async loadMember(member: Member): Promise<void> {
+    let assocs = await this.memberMonitorAssociationRepository.find({
+      where: {
+        memberId: member.id,
+      },
+    });
+
+    let monitorIds = assocs.map(({monitorId}) => monitorId);
+    member.monitors = await this.monitorRepository.find({
+      where: {
+        id: {
+          inq: monitorIds,
+        },
+      },
+    });
+  }
+
+  private async loadEndpointpolicy(policy: Endpointpolicy): Promise<void> {
+    policy.rules = await this.endpointpolicyRepository.rules(policy.id).find();
+
+    if (policy.rules.length === 0) {
+      throw new HttpErrors.UnprocessableEntity(
+        'Endpoint Policy ' + policy.id + ' has no rule',
+      );
+    }
+
+    for (let rule of policy.rules) {
+      await this.loadRule(rule);
+    }
+  }
+
+  private async loadRule(rule: Rule): Promise<void> {
+    rule.conditions = await this.ruleRepository.conditions(rule.id).find();
+    rule.actions = await this.ruleRepository.actions(rule.id).find();
+
+    for (let action of rule.actions) {
+      await this.loadAction(action);
+    }
+  }
+
+  private async loadAction(action: Action): Promise<void> {
+    if (action.type === 'waf' && action.policy) {
+      action.wafpolicy = await this.wafpolicyRepository.findById(action.policy);
+    }
+  }
+
   @post(prefix + '/applications/{applicationId}/declarations', {
```

Comment:
To my thinking, Declaration should have its own controller, instead of applications. 
Also: `'/applications/{applicationId}/declarations`: I think we needn't an application id to index declaration. 

The relation of declaration and application is 'related' not 'belong to'. 

Users can deploy the declaration, which already has no relation to the application. 

Think that if the user changes the application-tree again, and deploy an older declaration with API: '/applications/{applicationId}/declarations' will make users confused, application-tree is newer than declaration I am deploying.

Reply:
One declaration only belongs to one application. It can not be association with several applications.

Reply:
I didn't say that the declaration can be associated with multiple application. 

"One declaration only belongs to one application." I may not agree with this statement. 

it's not a belong-to relation, but a related. In fact, the declaration is related with not only application but kinds of other configurations. 

The declaration should have its own controller/model, also endpoint prefix. I think. That is my core thinking/main concern.

Reply:
In our resource model, this 'declaration' terminology just mean the declaration of an application. We will not create or deploy any other kind of AS3 declaration.

Reply:
Got it.